library/std/src/sys_common/lazy_box.rs

```diff
@@ -0,0 +1,66 @@
+#![allow(dead_code)] // Only used on some platforms.
+
+// This is used to wrap pthread {Mutex, Condvar, RwLock} in.
+
+use crate::marker::PhantomData;
+use crate::ops::{Deref, DerefMut};
+use crate::ptr::null_mut;
+use crate::sync::atomic::{
+    AtomicPtr,
+    Ordering::{AcqRel, Acquire},
+};
+
+pub(crate) struct LazyBox<T: LazyInit> {
+    ptr: AtomicPtr<T>,
+    _phantom: PhantomData<T>,
+}
+
+pub(crate) unsafe trait LazyInit {
+    fn new() -> Self;
```

Comment:
This should document that `new` and `init` might be called multiple times when threads race to `get_pointer` on the lazy box.